File: FBSimulatorControl/Commands/FBSimulatorDebuggerCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorDebuggerCommands.h"

#import "FBSimulator.h"

@interface FBSimulatorDebugServer : NSObject <FBDebugServer>

@property (nonatomic, strong, readonly) FBProcess<NSNull *, id<FBControlCoreLogger>, id<FBControlCoreLogger>> *task;

@end

@implementation FBSimulatorDebugServer

@synthesize lldbBootstrapCommands = _lldbBootstrapCommands;
@synthesize completed = _completed;

- (instancetype)initWithDebugServerTask:(FBProcess<NSNull *, id<FBControlCoreLogger>, id<FBControlCoreLogger>> *)task lldbBootstrapCommands:(NSArray<NSString *> *)lldbBootstrapCommands
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _task = task;
  _lldbBootstrapCommands = lldbBootstrapCommands;

  return self;
}

#pragma mark FBiOSTargetOperation

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunused-property-ivar"

- (FBFuture<NSNull *> *)completed
{
  FBProcess *task = self.task;
  return [[[task
    statLoc]
    mapReplace:NSNull.null]
    onQueue:dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0) respondToCancellation:^{
      return [task sendSignal:SIGTERM backingOffToKillWithTimeout:1 logger:nil];
    }];
}

#pragma clang diagnostic pop

@end

@interface FBSimulatorDebuggerCommands ()

@property (nonatomic, weak, readonly) FBSimulator *simulator;
@property (nonatomic, copy, readonly) NSString *debugServerPath;

@end

@implementation FBSimulatorDebuggerCommands

#pragma mark Initializers

+ (NSString *)debugServerPath
{
  return [[FBXcodeConfiguration
    contentsDirectory]
    stringByAppendingPathComponent:@"SharedFrameworks/LLDB.framework/Resources/debugserver"];
}

+ (instancetype)commandsWithTarget:(FBSimulator *)target
{
  return [[self alloc] initWithSimulator:target debugServerPath:[self debugServerPath]];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator debugServerPath:(NSString *)debugServerPath
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;
  _debugServerPath = debugServerPath;

  return self;
}

#pragma mark FBDebuggerCommands

- (FBFuture<id<FBDebugServer>> *)launchDebugServerForHostApplication:(FBBundleDescriptor *)application port:(in_port_t)port
{
  FBApplicationLaunchConfiguration *configuration = [[FBApplicationLaunchConfiguration alloc]
    initWithBundleID:application.identifier
    bundleName:application.name
    arguments:@[]
    environment:@{}
    waitForDebugger:YES
    io:FBProcessIO.outputToDevNull
    launchMode:FBApplicationLaunchModeFailIfRunning];
  return [[[self.simulator
    launchApplication:configuration]
    onQueue:self.simulator.workQueue fmap:^(id<FBLaunchedApplication> process) {
      return [self debugServerTaskForPort:port processIdentifier:process.processIdentifier];
    }]
    onQueue:self.simulator.workQueue map:^(FBProcess<NSNull *, id<FBControlCoreLogger>, id<FBControlCoreLogger>> *task) {
      NSArray<NSString *> *lldbBootstrapCommands = @[
        [NSString stringWithFormat:@"process connect connect://localhost:%d", port]
      ];
      return [[FBSimulatorDebugServer alloc] initWithDebugServerTask:task lldbBootstrapCommands:lldbBootstrapCommands];
    }];
}

#pragma mark Private

- (FBFuture<FBProcess<NSNull *, id<FBControlCoreLogger>, id<FBControlCoreLogger>> *> *)debugServerTaskForPort:(in_port_t)port processIdentifier:(pid_t)processIdentifier
{
  return [[[[[FBProcessBuilder
    withLaunchPath:self.debugServerPath]
    withArguments:@[[NSString stringWithFormat:@"localhost:%d", port], @"--attach", [NSString stringWithFormat:@"%d", processIdentifier]]]
    withStdOutToLogger:self.simulator.logger]
    withStdErrToLogger:self.simulator.logger]
    start];
}

@end

File: FBSimulatorControl/Commands/FBSimulatorDebuggerCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

// MARK: - FBSimulatorDebugServer

private class FBSimulatorDebugServer: NSObject, FBDebugServer {

  let task: FBSubprocess<NSNull, AnyObject, AnyObject>
  let lldbBootstrapCommands: [String]

  init(debugServerTask: FBSubprocess<NSNull, AnyObject, AnyObject>, lldbBootstrapCommands: [String]) {
    self.task = debugServerTask
    self.lldbBootstrapCommands = lldbBootstrapCommands
    super.init()
  }

  // MARK: - FBiOSTargetOperation

  var completed: FBFuture<NSNull> {
    let task = self.task
    let queue = DispatchQueue.global(qos: .userInitiated)

    // Resolve `mutable` when statLoc completes, but observe via
    // `notifyOfCompletion` instead of `mapReplace`. `mapReplace` wires up an
    // automatic respondToCancellation that cancels the source future when the
    // chain is cancelled. That auto-responder races with our SIGTERM responder
    // below: if it wins, `task.statLoc` becomes cancelled (hasCompleted = YES),
    // and `task.sendSignal:`'s "skip if dead" guard short-circuits without
    // calling `kill()`. Net effect: the process is never signaled.
    let mutable = FBMutableFuture<NSNull>()
    task.statLoc.onQueue(
      queue,
      notifyOfCompletion: { _ in
        mutable.resolve(withResult: NSNull())
      })

    return convertFBMutableFuture(mutable).onQueue(
      queue,
      respondToCancellation: {
        // sendSignal returns FBFuture<NSNumber>. Map to NSNull so the
        // responder's future actually matches its declared return type and
        // the bridge can read its result safely.
        task.sendSignal(SIGTERM, backingOffToKillWithTimeout: 1, logger: nil)
          // swiftlint:disable:next force_cast
          .mapReplace(NSNull()) as! FBFuture<NSNull>
      }
    )
  }
}

// MARK: - FBSimulatorDebuggerCommands

public final class FBSimulatorDebuggerCommands: NSObject, FBiOSTargetCommand {

  // MARK: - Properties

  internal weak var simulator: FBSimulator?
  internal let debugServerPath: String

  // MARK: - Class Methods

  internal class func resolveDebugServerPath() -> String {
    (FBXcodeConfiguration.contentsDirectory as NSString)
      .appendingPathComponent("SharedFrameworks/LLDB.framework/Resources/debugserver")
  }

  // MARK: - Initializers

  public class func commands(with target: any FBiOSTarget) -> FBSimulatorDebuggerCommands {
    FBSimulatorDebuggerCommands(
      // swiftlint:disable:next force_cast
      simulator: target as! FBSimulator,
      debugServerPath: resolveDebugServerPath()
    )
  }

  internal init(simulator: FBSimulator, debugServerPath: String) {
    self.simulator = simulator
    self.debugServerPath = debugServerPath
    super.init()
  }

  // MARK: - Private

  func launchDebugServer(forHostApplication application: FBBundleDescriptor, port: in_port_t) async throws -> any FBDebugServer {
    guard let simulator = self.simulator else {
      throw FBSimulatorError.describe("Simulator deallocated").build()
    }
    let configuration = FBApplicationLaunchConfiguration(
      bundleID: application.identifier,
      bundleName: application.name,
      arguments: [],
      environment: [:],
      waitForDebugger: true,
      io: FBProcessIO<AnyObject, AnyObject, AnyObject>.outputToDevNull(),
      launchMode: .failIfRunning
    )
    let launchedApp = try await simulator.launchApplication(configuration)
    // swiftlint:disable:next force_cast
    let debugTask = try await bridgeFBFuture(debugServerTask(forPort: port, processIdentifier: launchedApp.processIdentifier, simulator: simulator, debugServerPath: debugServerPath)) as! FBSubprocess<NSNull, AnyObject, AnyObject>
    let lldbBootstrapCommands = [
      "process connect connect://localhost:\(port)"
    ]
    return FBSimulatorDebugServer(
      debugServerTask: debugTask,
      lldbBootstrapCommands: lldbBootstrapCommands
    )
  }

  private func debugServerTask(forPort port: in_port_t, processIdentifier: pid_t, simulator: FBSimulator, debugServerPath: String) -> FBFuture<AnyObject> {
    FBProcessBuilder<NSNull, AnyObject, AnyObject>
      .withLaunchPath(debugServerPath)
      .withArguments(["localhost:\(port)", "--attach", "\(processIdentifier)"])
      // swiftlint:disable:next force_unwrapping
      .withStdOut(to: simulator.logger!)
      // swiftlint:disable:next force_unwrapping
      .withStdErr(to: simulator.logger!)
      // swiftlint:disable:next force_cast
      .start() as! FBFuture<AnyObject>
  }
}

// MARK: - FBSimulator+DebuggerCommands

extension FBSimulator: DebuggerCommands {

  public func launchDebugServer(
    forHostApplication application: FBBundleDescriptor,
    port: in_port_t
  ) async throws -> any FBDebugServer {
    try await debuggerCommands().launchDebugServer(forHostApplication: application, port: port)
  }
}
